### PR TITLE
jsonschema: configure the holding editor

### PR DESCRIPTION
**Pull request by jma**
The holding JSONSchema is adapted to be usable by the admin ui editor.

* Fixes permissions in the API REST serializer: create is renamed to
cannot_create.
* Fixes the holding importation by defining a default value for the
holding harvested properties.
* Fixes the returned http status when wrong data are posted to the api
holding views.
* Generates documents and persons data after the MEF reindexing.
* Generates holdings and items data.

Co-Authored-by: Johnny Mariéthoz <Johnny.Mariethoz@rero.ch>
Co-Authored-by: Aly Badr <aly.badr@rero.ch>

## Why are you opening this PR?

- Implements: https://tree.taiga.io/project/rero21-reroils/us/1039?milestone=260625

## Dependencies

- rero-ils-ui: https://github.com/rero/rero-ils-ui/pull/216

## How to test?

- run the usual bootstrap and setup commands ( new data has been generated for this PR )
- check holding and item creation for the 3 types of holdings
  - journals: should display a holding editor
  - monography: should display a item editor
  - e-books: no editor should be available
- here is a list of documents with existing serials holdings
  - https://ilsdev.test.rero.ch/professional/records/documents/detail/152
  - https://ilsdev.test.rero.ch/professional/records/documents/detail/1101
  - https://ilsdev.test.rero.ch/professional/records/documents/detail/42
  - https://ilsdev.test.rero.ch/professional/records/documents/detail/617
  - https://ilsdev.test.rero.ch/professional/records/documents/detail/70
  - https://ilsdev.test.rero.ch/professional/records/documents/detail/689
  - https://ilsdev.test.rero.ch/professional/records/documents/detail/1148
  - https://ilsdev.test.rero.ch/professional/records/documents/detail/1089
  - https://ilsdev.test.rero.ch/professional/records/documents/detail/866
  - https://ilsdev.test.rero.ch/professional/records/documents/detail/597
  - https://ilsdev.test.rero.ch/professional/records/documents/detail/1184
- play with the holding editor check the titles and descriptions messages

## Code review check list

- [x] Commit message template compliance.
- [x] Commit message without typos.
- [x] File names.
- [x] Functions names.
- [x] Functions docstrings.
- [x] Unnecessary commited files?
- [ ] Extracted translations?

### CODE REVIEW

File: rero_ils/modules/holdings/jsonschemas/holdings/holding-v0.0.1.json [hunk 1]
```diff
@@ -92,116 +113,167 @@
...
       "required": [
         "values",
         "template"
       ],
       "additionalProperties": false,
       "properties": {
         "template": {
-          "title": "Template for the display of the holdings issues.",
+          "title": "Issue Display Template",
+          "description": "This template define how a prediction is displayed. i.e. year 2020; vol.1, iss. 4. It should contains the variable field between {{}} with the corresponding pattern and level name: i.e. {{chronology.year}}; vol. {{enumeration.volume}}, iss. {{enumeration.issue}}.",
```
Comment (BadrAly) [inline rero_ils/modules/holdings/jsonschemas/holdings/holding-v0.0.1.json:147]: replace `This template define how a prediction is displayed` by `This template define how an issue is displayed` ?

**Comment by JoelleDosimont** [comment]
Here are my comments : 
•	bug : as Nicolas stated, there is a bug with the use of the "list based prediction". It's ok when creating, but once saved, it changes to "Number based". Example : https://ilsdev.test.rero.ch/professional/records/holdings/edit/6468 
•	Error : the system shouldn't allow to save a prediction when an error is found. See : https://ilsdev.test.rero.ch/professional/records/holdings/edit/6471 
•	When I use the "list based", I have to manually add all the month in the value. It would be nice to have a defined list (but I realize it's difficult to adapt it by language). 
•	The preview is really helpful to construct the prediction and see if everything goes fine.  Ideally, it should adapt at the count of the "completion value" to be sure the change in enumeration is ok. Example : https://ilsdev.test.rero.ch/professional/records/holdings/edit/6470  (as "only" the first 10 number are displayed, I cannot see if everything is good, as the change occurs after 24 numbers).  
•	How can I combine some numbers. Example : numbers for July and August are published each year together. 
•	Can I use ordinal numbers : 1st issue, 2nd issue, etc. ? 
•	If the prediction changes, what should I do ? 
•	I'm not sure it’s a good idea to force the name of the chronology field to always be "Level_1 / Level_2…". In my understanding, when it's about chronology, there should only be : day, week, month, year, season. As for enumeration there can be "level_1/2/3/4…

At first, it's difficult to know how it works, but when it's understood it really is flexible and it's a question of logic. I would also enjoy to find some basic template to helps for commun predictions.

Nice work !!

**Comment by jma** [comment]
> * Periodicity: we miss this element from the prediction (as discussed with @BadrAly ). Okay, it is only a pattern. But very soon, we will need it to know _when in time_ the next issue is expected.

This will be addressed in the next US as discussed at the grooming.

> * **Bug**
>   
>   * when I edit a holding with a pattern with a list, it does not display my list items that I entered manually. Try to edit this one to reproduce it: https://ilsdev.test.rero.ch/api/holdings/6409

Solved.

> * Editor
>   
>   * Name of the fields and of the values in enumerations: only first letter must be uppercase
>   * `Cote de l'état de collection` > simplify: `Cote`
Done.
> * The name of the chronology levels
>   
>   * I would place it as the 1st field of the chronology level, because it is a sort of _title_ of it (so: above the type)
>   * It makes it very hard to distinguish between _names of chronology levels_ and _names of patterns_. I would allow only to choose names of patterns. As the chronology levels are always in hierarchical relation, I would force their names, and make them not editable. Ex: level_1, level_2

Did nothing, we need an agreement for this. I propose to make the modification after the feedback of the pilot libraries.

> * Descriptions
>   
>   * Field _Issue display template_, shorten the description. Insert the examples of how to enter a template as a placeholder, according to our [usability charter](https://github.com/rero/rero-ils/wiki/Usability-charter). So in placeholder: _Example: Vol. {{numbering.vol}}, {{time.year}}_
>   * Field _Name_ of the prediction pattern, remove "Name of the pattern" from description, it is useless.
>   * Field _Name_ of the chronology level, remove "Name of the pattern level." from description, it is useless.
>   * Field _Starting Value_, remove "Default 1". It is obvious, must not be explicitly announced to end user.
>   * _Type_ of chronology level (either number based or list based), add a description to explain the difference: _List based predictions are used for seasons, months or other chronologies that are not number based._

Done.

> * Default values: at creation of a new prediction pattern, I would insert default values for a very common prediction case. Because starting from scratch is almost impossible.

The first approach is to use a pre-fill editor as discussed with Nicolas. In a second step we will propose the use of models. This will be available for all kind of resources: documents, holdings, etc. This should be an US as discussed few weeks ago.

> * Usability problem
>   
>   * Next predicted value: in some cases, we must modify this value at various places/levels for the prediction to work well.

Too complicated for now.

> 
> General comment: for now, the form is complex and difficult to understand. But the preview works very well and helps a lot to create a prediction. It almost makes it playful. This is a great improvement!

**Comment by jma** [comment]
> Here are my comments :
> • bug : as Nicolas stated, there is a bug with the use of the "list based prediction". It's ok when creating, but once saved, it changes to "Number based". Example : https://ilsdev.test.rero.ch/professional/records/holdings/edit/6468

Done.

> • Error : the system shouldn't allow to save a prediction when an error is found. See : https://ilsdev.test.rero.ch/professional/records/holdings/edit/6471

Can you add an issue for this?

> • When I use the "list based", I have to manually add all the month in the value. It would be nice to have a defined list (but I realize it's difficult to adapt it by language).

Will be solved with the use of models.

> • The preview is really helpful to construct the prediction and see if everything goes fine. Ideally, it should adapt at the count of the "completion value" to be sure the change in enumeration is ok. Example : https://ilsdev.test.rero.ch/professional/records/holdings/edit/6470 (as "only" the first 10 number are displayed, I cannot see if everything is good, as the change occurs after 24 numbers).

It is difficult to do automatically I propose to add the possibility to change the number of preview examples. This can be added to the next US concerning the serials.

> • How can I combine some numbers. Example : numbers for July and August are published each year together.

Using a list based chronology level.

> • Can I use ordinal numbers : 1st issue, 2nd issue, etc. ?

Can you add an issue for this?

> • If the prediction changes, what should I do ?

This is not clear for now, but probably nothing. The prediction will be used only for the new items.

> • I'm not sure it’s a good idea to force the name of the chronology field to always be "Level_1 / Level_2…". In my understanding, when it's about chronology, there should only be : day, week, month, year, season. As for enumeration there can be "level_1/2/3/4…

Me too.

> 
> At first, it's difficult to know how it works, but when it's understood it really is flexible and it's a question of logic. I would also enjoy to find some basic template to helps for commun predictions.
> 
> Nice work !!
 Thanks.

**Pull request merged**